net/multi_listen_test.go

```diff
@@ -0,0 +1,282 @@
+package net
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+type fakeCon struct {
+	remoteAddr net.Addr
+}
+
+func (f *fakeCon) Read(_ []byte) (n int, err error) {
+	return 0, nil
+}
+
+func (f *fakeCon) Write(_ []byte) (n int, err error) {
+	return 0, nil
+}
+
+func (f *fakeCon) Close() error {
+	return nil
+}
+
+func (f *fakeCon) LocalAddr() net.Addr {
+	return nil
+}
+
+func (f *fakeCon) RemoteAddr() net.Addr {
+	return f.remoteAddr
+}
+
+func (f *fakeCon) SetDeadline(_ time.Time) error {
+	return nil
+}
+
+func (f *fakeCon) SetReadDeadline(_ time.Time) error {
+	return nil
+}
+
+func (f *fakeCon) SetWriteDeadline(_ time.Time) error {
+	return nil
+}
+
+var _ net.Conn = &fakeCon{}
+
+type fakeListener struct {
+	addr         net.Addr
+	index        int
+	err          error
+	closed       atomic.Bool
+	connErrPairs []connErrPair
+}
+
+func (f *fakeListener) Accept() (net.Conn, error) {
+	if f.index < len(f.connErrPairs) {
+		index := f.index
+		connErr := f.connErrPairs[index]
+		f.index++
+		return connErr.conn, connErr.err
+	}
+	for {
+		if f.closed.Load() {
+			return nil, fmt.Errorf("use of closed network connection")
+		}
+	}
+}
+
+func (f *fakeListener) Close() error {
+	f.closed.Store(true)
+	return nil
+}
+
+func (f *fakeListener) Addr() net.Addr {
+	return f.addr
+}
+
+var _ net.Listener = &fakeListener{}
+
+func TestMultiListen(t *testing.T) {
```

Comment:
I think we need some tests of accept-while-close, close-with-pending-connections, etc

Reply:
and some tests of HTTP serving on a multi-listener

Reply:
> and some tests of HTTP serving on a multi-listener

do you mean an integration test that uses real HTTP servers?

Reply:
I think it can still be a unit test (since this repo doesn't really have integration tests).  Just use `:0` for a random port.  Create 2 or 3 such ports, pass the multi-listener to `http.Serve` and then connect to each of the ports that were allocated, ensuring they get the correct HTTP response.